### PR TITLE
[WIP] Integration tests

### PR DESCRIPTION
**Pull request by girdharshubham**
# Changelog
* build.sbt
* AllApis
* IntegrationSpecHelper
* IntegrationSpec

### CODE REVIEW

File: zio-http/src/it/scala/zhttp/AllApis.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  def api = HttpApp.collect[Any, Nothing] {
+    case Method.GET -> !!                   => Response.ok
+    case Method.POST -> !!                  => Response.status(Status.CREATED)
+    case Method.GET -> !! / "continue"      => Response.ok
+    case Method.GET -> !! / "boom"          => Response.status(Status.INTERNAL_SERVER_ERROR)
+    case Method.GET -> !! / "subscriptions" => Response.socket(socket)
+  }
+
+  def apiM = HttpApp.collectM[Clock, Nothing] { case Method.GET -> !! / "timeout" =>
```
Comment (amitksingh1490) [inline zio-http/src/it/scala/zhttp/AllApis.scala]: this is not needed as we will add a test for timeout middleware once that is merged

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  import IntegrationSpecHelper._
+
+  def spec = suite("IntegrationSpec")(
+    ApiSpec,
+  ).provideCustomLayer(env) @@ timeout(10 seconds)
+
+  def ApiSpec = suite("ApiSpec") {
+    testM("200 ok on /") {
+      val zResponse = Client.request(baseAddr)
+      assertM(zResponse.map(_.headerRemoved("date")))(
```
Comment (amitksingh1490) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:23]:
```suggestion
      assertM(zResponse)(
```

We can do this and make servertime configurable

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        equalTo(
+          HttpResponse(
+            Status.OK,
+            List(Header.custom("server", "ZIO-Http"), Header.custom("content-length", "0")),
+            CompleteData(Chunk.empty),
```
Comment (amitksingh1490) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala]: in release/next we have a assertion named isResponse and you can check response.status res.headers etc. may be we can move those to reduce the verbosity

Comment (amitksingh1490) [reply zio-http/src/it/scala/zhttp/IntegrationSpec.scala]: also in this case if HttpData.CompleteData is renamed all tests needs to be changed however if you have a getContent function if something changes in the API you only need to change getContent

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      } +
+      testM("500 internal server error on /boom") {
+        val zResponse = Client.request((Method.GET, "/boom".url))
+        assertM(zResponse.getStatus)(status(INTERNAL_SERVER_ERROR))
+      } +
+      testM("200 ok on Stream file") {
+        val zResponse = Client.request((Method.GET, "/stream/file".url))
+        assertM(zResponse.getStatus)(status(OK))
+      } @@ ignore
+  }
```
Comment (amitksingh1490) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:61]: add tests for Headers and content also

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  Runtime.default.unsafeRun(server.make.useForever.provideSomeLayer(env).forkDaemon)
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:63]: Remove `unsafeRun`. The test suite will use it internally.

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  def env    = EventLoopGroup
+    .auto() ++ ChannelFactory.auto ++ ServerChannelFactory.auto ++ zio.clock.Clock.live ++ zio.blocking.Blocking.live
+
+  def spec = suite("IntegrationSpec")(
+    StatusCodeSpec,
+  ).provideCustomLayer(env) @@ timeout(30 seconds)
+
+  def StatusCodeSpec = suite("StatusCodeSpec") {
+    testM("200 ok on /[GET]") {
+      val zResponse = Client.request((Method.GET, "/".url))
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:28]:
```suggestion
      val zResponse = request()
```

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+      assertM(zResponse.getStatus)(status(OK))
+    } +
+      testM("201 created on /[POST]") {
+        val zResponse = Client.request((Method.POST, "/".url))
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:32]:
```suggestion
        val zResponse = Client.request(Method.POST)
```

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        val zResponse = Client.request((Method.PUT, "/".url))
+        assertM(zResponse.getStatus)(status(NO_CONTENT))
+      } +
+      testM("204 no content ok on /[DELETE]") {
+        val zResponse = Client.request((Method.DELETE, "/".url))
+        assertM(zResponse.getStatus)(status(NO_CONTENT))
+      } +
+      testM("404 on a random route") {
+        checkM(Gen.int(1, 2).flatMap(Gen.stringN(_)(Gen.alphaNumericChar))) { randomUrl =>
+          val zResponse = Client.request((Method.GET, randomUrl.url))
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:45]:
```suggestion
          val zResponse = request(randomUrl)
```

File: zio-http/src/it/scala/zhttp/IntegrationSpec.scala [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+          assertM(zResponse.getStatus)(status(NOT_FOUND))
+        }
+      } +
+      testM("400 bad request on /subscriptions without the connection upgrade header") {
+        val zResponse = Client.request((Method.GET, "/subscriptions".url))
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/IntegrationSpec.scala:50]:
```suggestion
        val zResponse = request(!! / "subscriptions")
```

File: zio-http/src/it/scala/zhttp/internal/AllApis.scala [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+  def app = HttpApp.collect {
+    case Method.GET -> !!                     => Response.ok
+    case Method.POST -> !!                    => Response.status(Status.CREATED)
+    case Method.PUT -> !!                     => Response.status(Status.NO_CONTENT)
+    case Method.DELETE -> !!                  => Response.status(Status.NO_CONTENT)
+    case Method.GET -> !! / "boom"            => Response.status(Status.INTERNAL_SERVER_ERROR)
+    case Method.GET -> !! / "subscriptions"   => Response.socket(socket)
+    case Method.GET -> !! / "stream" / "file" => Response.http(content = file)
+  }
```
Comment (tusharmath) [inline zio-http/src/it/scala/zhttp/internal/AllApis.scala:24]:
```scala
type Id

case class DynamicHttpServer[R, E](ref: Ref[List[HttpApp[R, E]]]) {
  def put(app: HttpApp[R, E]): UIO[Id] = ???
  def delete(app: Id): UIO[Boolean] = ???
  def get(id: Id): UIO[HttpApp[R, E]] = ???
}

val app = Http.collect {
  case req @ _ -> id /: path !! => dServer.get(id).flatMap(app => app(req))
}
.flatten

Server.make(app).useForever

test("example") {
  val app = dServer.put(Http.ok)
  val response = client.request(URL ++ app.id)
  assertM(response)( ... )
}

```

**Comment by tusharmath** [comment]
changes taken in #575

**Pull request closed without merging**